File: 1020-number-of-enclaves/1020-number-of-enclaves.py
class Solution:
    def dfs(self,grid,r,c,m,n,directions):
        if(r>=0 and r<m and c>=0 and c<n and grid[r][c]==1):
            grid[r][c]=0
            for dr,dc in directions:
                self.dfs(grid,dr+r,dc+c,m,n,directions)
                
    def numEnclaves(self, grid: List[List[int]]) -> int:
        m=len(grid)
        n=len(grid[0])
        directions=[[1,0],[-1,0],[0,1],[0,-1]]
        for r in range(m):
            for c in range(n):
                if r==0 or c==0 or r==m-1 or c==n-1:
                    self.dfs(grid,r,c,m,n,directions)
        ans=0
        for r in range(m):
            for c in range(n):
                if grid[r][c]==1:
                    ans+=1
        return ans            

File: 1020-number-of-enclaves/1020-number-of-enclaves.cpp
class Solution {
public:
    int m;
    int n;
    int dr[4]={-1,0,1,0};
    int dc[4]={0,1,0,-1};
    void dfs(vector<vector<int>> &grid,int r,int c){
        if(r>=0 and r<m and c>=0 and c<n and grid[r][c]==1){
            grid[r][c]=0;
            for(int i=0;i<4;i++){
                dfs(grid,r+dr[i],c+dc[i]);
            }
        }
    }
    int numEnclaves(vector<vector<int>>& grid) {
        m=grid.size();
        n=grid[0].size();
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if((i==0 || i==m-1 || j==0 || j==n-1) && grid[i][j]==1 ){
                    dfs(grid,i,j);
                }
            }
        }
        int ans=0;
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(grid[i][j]==1){
                    ans++;
                }
            }
        }
        return ans;
        
    }
};